codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -253,9 +253,8 @@ public final void userEventTriggered(final ChannelHandlerContext ctx, final Obje
             // handlerAdded(...) which will be run before other handlers will be added to the pipeline.
             ctx.executor().execute(() -> ctx.fireUserEventTriggered(evt));
         } else if (evt instanceof UpgradeEvent) {
-            UpgradeEvent upgrade = (UpgradeEvent) evt;
-            try {
-                onUpgradeEvent(ctx, upgrade.retain());
+            try (UpgradeEvent upgrade = (UpgradeEvent) evt) {
+                onUpgradeEvent(ctx, upgrade);
```

Comment:
Hmmm .. yeah this is wrong, it seems I will have to copy the event data because as you said `onUpgradeEvent` takes ownership of the event but we need it afterwards below too. This sucks as we don't even know if anyone will use this event/request